content/2015.09.kinto-encryption-example.rst

```diff
@@ -0,0 +1,207 @@
+This blogpost shows how to use a <code>RemoteTransformer</code> for encrypting your data just
+before it's sent to the cloud. It uses WebCrypto for the cryptography.
+
+The main idea behind a <code>RemoteTransformer</code> is that you provide two functions,
+"encode" and "decode", and Kinto.js makes sure to call the encode function on
+each record that gets uploaded, and to call the decode function on each record
+that gets downloaded. You can also read about this in the Kinto.js documentation,
+http://kintojs.readthedocs.org/.
+
+NB: I am not a security expert and I wrote this example in an afternoon, so
+if you're going to copy-paste this code, please don't use it for anything
+valuable without doing your own proper security review first! :)
+
+The way we tell Kinto.js that we want to use encryption, and provide our own
+"encode" and "decode" procedures, is quite simple. When creating a Kinto
+collection, we pass a <code>RemoteTransformer</code> as a collection-level option. The full
+code example is
+`on github http://michielbdejong.github.io/kinto-encryption-example/app.js`_,
+but let's walk through the code step-by-step.
+
+If you're unfamiliar with ES6, you may be surprised by some of this syntax; so let's
+take this opportunity to get used to it! Here's a little translation:
+* <code>() => {}</code> is similar to <code>function() {}</code>
+(although the scoping of <code>this</code> is different).
+* <code>const foo = 'bar'</code> and <code>let foo = 'bar'</code> are similar to
+<code>var foo = 'bar'</code>.
+* <code>{ a, b }</code> is similar to <code>{ a: a, b: b }</code>.
+* <code>`Hello ${foo}`</code> (note the backticks) is similar to
+<code>'Hello ' + foo</code>.
+
+First, we need a few helper functions, which will be used later on to convert
+between binary buffers and ASCII strings.
+
+.. code-block:: javascript
+  // Helper functions:
+  const rawStringToByteArray = (str) => {
+    const strLen = str.length;
+    var byteArray = new Uint8Array(strLen);
+    for (var i = 0; i < strLen; i++) {
+      byteArray[i] = str.charCodeAt(i);
+    }
+    return byteArray;
+  };
+  const base64StringToByteArray = (base64) => {
+    return rawStringToByteArray(window.atob(base64));
+  };
+  const byteArrayToBase64String = (buffer) => {
+    var bytes = new Uint8Array(buffer);
+    var binary = '';
+    for (var i=0; i<bytes.byteLength; i++) {
+        binary += String.fromCharCode(bytes[i]);
+    }
+    return window.btoa(binary);
+  };
+
+Next, we need to create a <code>RemoteTransformer</code> that can encode and decode. And for
+that, we need to generate an AES key. We will use that key to encrypt and decrypt
+the data that is passed through the <code>RemoteTransformer</code>:
+
+.. code-block:: javascript
+  // RemoteTransformer:
+  const generateAesKey = () => {
+    // See http://www.w3.org/TR/WebCryptoAPI/#examples-symmetric-encryption
+    return window.crypto.subtle.generateKey({ name: 'AES-CBC', length: 128 },
```

Comment:
I feel like introducing what the WebCrypto API is to the reader before using it in code snippets could help.